In [35]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# LIBRERIAS SKLEARN
from sklearn.model_selection import train_test_split

# LIBRERIA DE SKLEARN PARA GENERAR ARBOLES DE DECISION
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import plot_tree
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

# LIBRERIA PLOTLY PARA GRAFICAR LOS RESULTADOS
import plotly.express as px
import plotly.graph_objects as go

In [36]:
df = pd.read_csv("../DATASETS/df_enfermedades_cardiacas.csv")
df

,Tipo_Dolor_Pecho,Frecuencia_Cardiaca_Maxima,Angina_Ejercicio,Depresion_ST_Ejercicio,Pendiente_Segmento_ST,Numero_Vasos_Filtrados,Enfermedad_Cardiaca
0,0,168,0,1.0,2,2,0
1,0,155,1,3.1,0,0,0
2,0,125,1,2.6,0,0,0
3,0,161,0,0.0,2,1,0
4,0,106,0,1.9,1,3,0
...,...,...,...,...,...,...,...
1020,1,164,1,0.0,2,0,1
1021,0,141,1,2.8,1,1,0
1022,0,118,1,1.0,1,1,0
1023,0,159,0,0.0,2,0,1


**SELECCION DE VARIABLES PARA ENTRENAR EL MODELO**

In [37]:
X = df.drop("Enfermedad_Cardiaca", axis=1)
y = df["Enfermedad_Cardiaca"]

# IMPRIMIR LAS 10 PRIMERAS FILAS DE X
print("--------------------------------------------------------")
print("X")
print(X.head(5))
print("--------------------------------------------------------")

# IMPRIMIR LAS 10 PRIMERAS FILAS DE Y
print("--------------------------------------------------------")
print("Y")
print("--------------------------------------------------------")
print(y.head(10))
print("--------------------------------------------------------")

--------------------------------------------------------
X
   Tipo_Dolor_Pecho  Frecuencia_Cardiaca_Maxima  Angina_Ejercicio  \
0                 0                         168                 0   
1                 0                         155                 1   
2                 0                         125                 1   
3                 0                         161                 0   
4                 0                         106                 0   

   Depresion_ST_Ejercicio  Pendiente_Segmento_ST  Numero_Vasos_Filtrados  
0                     1.0                      2                       2  
1                     3.1                      0                       0  
2                     2.6                      0                       0  
3                     0.0                      2                       1  
4                     1.9                      1                       3  
--------------------------------------------------------
-------------------

**ENTRENAMIENTO Y PRUEBA DEL MODELO**

In [38]:
# SE VA A UTILIZAR EL 80% DE LOS DATOS PARA ENTRENAR EL MODELO Y EL 20% PARA PROBARLO
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [39]:
# SE PROCEDE A INSTANCIAR EL MODELO DE ARBOL DE DECISION
arbol_decision = DecisionTreeClassifier()

In [40]:
arbol_decision.fit(X_train, y_train)

# SE CREA LA VARIABLE DE PREDICCION
y_pred = arbol_decision.predict(X_test)
y_pred

array([1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0,
       1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0,
       0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1,
       0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0,
       1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1,
       0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1,
       1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0,
       1, 1, 1, 1, 1, 0, 0])

In [41]:
# CALCULAMOS LA METRICAS (ACURACY, PRECISION, RECALL Y F1-SCORE)

from sklearn.metrics import precision_score, recall_score, f1_score

acuraccy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
f1_score = f1_score(y_test, y_pred)

print("--------------------------------------------------------")
print("El valor Acuraccy es: {}".format(acuraccy))
print("El valor Precision es: {}".format(precision))
print("El valor F1-Score es: {}".format(f1_score))

--------------------------------------------------------
El valor Acuraccy es: 0.9853658536585366
El valor Precision es: 1.0
El valor F1-Score es: 0.9852216748768473


In [42]:
# SE IMPORTA LA MATRIZ DE CONFUSION Y REPORTE DE CLASIFICACION
from sklearn.metrics import classification_report, confusion_matrix

reporte_clasificacion = classification_report(y_test, y_pred)
matriz_confusion = confusion_matrix(y_test, y_pred)

print(reporte_clasificacion)

              precision    recall  f1-score   support

           0       0.97      1.00      0.99       102
           1       1.00      0.97      0.99       103

    accuracy                           0.99       205
   macro avg       0.99      0.99      0.99       205
weighted avg       0.99      0.99      0.99       205



**CONCLUSION:** Las métricas obtenidas evidencian un excelente desempeño del modelo de clasificación. Se alcanzó una exactitud (Accuracy) del 99%, lo que indica que la gran mayoría de las predicciones fueron correctas. Para la clase 0 (sin enfermedad), el modelo obtuvo un Recall de 100%, identificando correctamente todos los pacientes sanos presentes en el conjunto de prueba. Por su parte, para la clase 1 (con enfermedad), alcanzó un Recall de 97%, detectando correctamente la mayoría de los pacientes enfermos. Además, los valores de Precision y F1-Score cercanos a 1.0 en ambas clases reflejan un equilibrio adecuado entre la capacidad de detección y la reducción de errores de clasificación, lo que sugiere que el modelo es altamente confiable para este problema.

In [43]:
# SE GRAFICA LA MATRIZ DE CONFUSION EN PLOTLY
fig = go.Figure(data=go.Heatmap(z=matriz_confusion, x=["Predicho Negativo", "Predicho Positivo"], y=["Real Negativo", "Real Positivo"], colorscale="Viridis"))
fig.update_layout(title="Matriz de Confusión", xaxis_title="Predicción", yaxis_title="Real")
fig.show()

**CONCLUSION:** La matriz de confusión muestra que el modelo clasificó correctamente 102 pacientes sin enfermedad (verdaderos negativos) y 100 pacientes con enfermedad (verdaderos positivos). Además, no se presentaron falsos positivos, es decir, ningún paciente sano fue clasificado erróneamente como enfermo. Sin embargo, se registraron 3 falsos negativos, correspondientes a pacientes que realmente tenían la enfermedad pero fueron clasificados como sanos. En general, estos resultados evidencian una alta capacidad del modelo para distinguir entre ambas clases, con un número mínimo de errores de clasificación.

**EXPORTAR RESULTADOS**

In [44]:
# EXPORTANDO LOS RESULTADOS DE LAS PREDICCIONES A UN XLSX
resultados = pd.DataFrame({"Real": y_test, "Predicho": y_pred})
resultados["Comparacion"] = np.where(resultados["Real"] == resultados["Predicho"], "Correcta", "Incorrecta")
resultados.to_excel("../PREDICCIONES/COMPARACION_RANDOM_FOREST.xlsx", index=False)

print("Los resultados de las predicciones se han exportado correctamente a COMPARACION_RANDOM_FOREST.xlsx")

Los resultados de las predicciones se han exportado correctamente a COMPARACION_RANDOM_FOREST.xlsx
